jupyter_server/serverapp.py

```diff
@@ -1494,31 +1521,109 @@ def init_server_extension_config(self):
+                        app = extapp()
+                        app.link_to_serverapp(self)
+                        # Build a new list where we
+                        self._enabled_extensions[app.extension_name] = app
+                    elif extloc:
+                        extmod = importlib.import_module(extloc)
+                        func = _get_load_jupyter_server_extension(extmod)
+                        self._enabled_extensions[extloc] = extmod
+                    else:
+                        log_msg = _(
+                            "{module_name} is missing critical metadata. Check "
+                            "that _jupyter_server_extension_paths returns `app` "
+                            "and/or `module` as keys".format(module_name=module_name)
+                        )
+                        self.log.warn(log_msg)
+
+
+    def load_server_extensions(self):
         """Load any extensions specified by config.
 
         Import the module, then call the load_jupyter_server_extension function,
         if one exists.
 
         The extension API is experimental, and may change in future releases.
         """
-        # Initialize extensions
-        for modulename, enabled in sorted(self.jpserver_extensions.items()):
-            if enabled:
-                try:
-                    mod = importlib.import_module(modulename)
-                    func = getattr(mod, 'load_jupyter_server_extension', None)
-                    if func is not None:
-                        func(self)
-                        # Add debug log for loaded extensions.
-                        self.log.debug("%s is enabled and loaded." % modulename)
-                    else:
-                        self.log.warning("%s is enabled but no `load_jupyter_server_extension` function was found" % modulename)
-                except Exception:
-                    if self.reraise_server_extension_failures:
-                        raise
-                    self.log.warning(_("Error loading server extension %s"), modulename,
-                                  exc_info=True)
+
+        # Load all enabled extensions.
+        for extkey, extension in sorted(self._enabled_extensions.items()):
+            if isinstance(extension, ModuleType):
+                log_msg = (
+                    "Extension from {extloc} module enabled and "
+                    "loaded".format(extloc=extkey)
+                )
+            else:
+                log_msg = (
+                    "Extension {extension_name} enabled and "
+                    "loaded".format(extension_name=extension.extension_name)
+                )
+            # Find the extension loading function.
+            try:
+                # This function was prefixed with an underscore in in v1.0
+                # because this shouldn't be a public API for most extensions.
+                func = getattr(extension, '_load_jupyter_server_extension')
+                func(self)
```

Comment:
This is the line I meant should be guarded with an `"Error loading server extension"` warning.